Entity id translation service

escherize opened the pull request:
Resolves #47043

### What is this?

Embedding customers want to use entity ids. But our endpoints do not support looking things up by their entity ids (e.g. on `GET /api/card/:id`). This is a quick addition that adds an endpoint to translate from entity_id -> id.

In particular it adds an endpoint that takes a json body like 

```json
{"entity_ids": 
  {"card": ["entity_id1", "entity_id2"], 
   "dashboard" : ["entity_id3"]}}
```
where each key maps to a model, and the value is an array of entity ids

This endpoint returns a mapping from entity_id -> sequential id (aka id).

```json
{"entity_ids" {"entity_id1": {"id": 8, "status":"ok", "type":"card"},
               "entity_id2": {"id": 9, "status":"ok", "type":"card"}, 
               "entity_id3": {"id": 9, "status":"ok", "type":"dashboard"}}
```

The embedding user can then translate their requests and in effect use entity ids with out entire api.

### What's been done?

- [x] Tests for multiple entity ids
- [x] Tests for every table with an entity id
- [x] Test on api layer
- [x] Work-out shape of the api with FE consumer
- [x] Adds a standardized naming convention for models for FE <-> BE collaboration (reused in e.g. search too)
- [x] Perf testing

----

### Perf findings:
I've added 35,000 entities with entity_ids, and here are the results:

```clojure
  (def realistic-request
    (update-vals (:request (run-translation items))
                 (comp #(take 5 %) shuffle)))

  (-> realistic-request vals flatten count)
  ;; => 70
  
  (criterium/bench
   ;70 items, 35,000 entities total
   (api.embed.common/table->entity-ids->ids realistic-request))
   ;;   Evaluation count : 13260 in 60 samples of 221 calls.
   ;;              Execution time mean : 5.284078 ms
   ;;     Execution time std-deviation : 818.174829 µs
   ;;    Execution time lower quantile : 4.336682 ms ( 2.5%)
   ;;    Execution time upper quantile : 7.759709 ms (97.5%)
   ;;                    Overhead used : 1.898153 ns
   ;;
   ;; Found 4 outliers in 60 samples (6.6667 %)
   ;; 	low-severe	 2 (3.3333 %)
   ;; 	low-mild	 2 (3.3333 %)
   ;;  Variance from outliers : 85.8508 % Variance is severely inflated by outliers

  (def big-request
    (update-vals (:request (run-translation items))
                 (comp #(take 50 %) shuffle)))

  (-> big-request vals flatten count)

  (criterium/bench
   ;; => 700 items, 35,000 entities total
   (api.embed.common/table->entity-ids->ids big-request))
;; Evaluation count : 1260 in 60 samples of 21 calls.
;;              Execution time mean : 47.418405 ms
;;     Execution time std-deviation : 5.405070 ms
;;    Execution time lower quantile : 34.629211 ms ( 2.5%)
;;    Execution time upper quantile : 53.895442 ms (97.5%)
;;                    Overhead used : 1.898153 ns
;;
;; Found 1 outliers in 60 samples (1.6667 %)
;; 	low-severe	 1 (1.6667 %)
;;  Variance from outliers : 75.4886 % Variance is severely inflated by outliers
```

I think this is pretty fast already, with no optimizations. I'm not 100% sure what a typical request will look like, but with 70 items, we see ~5ms response time, and with 700 (much higher than usual, IME) we see ~50ms response time.

The current strategy is naive: we sequentially go through each table -> ids keypair passed into the handler, and find the [entity_id id] pairs for each. Finally we combine them into a map and return it.

If needed we can speed this up by parallelizing the lookups, caching the ids, and more. But it feels like a premature optimization at this time.

escherize committed:
entity id translation + tests

escherize committed:
Merge branch 'master' into bcm-entity-id-translation-service

escherize committed:
add api level test

escherize committed:
Merge branch 'bcm-entity-id-translation-service' of github.com:metabase/metabase into bcm-entity-id-translation-service

escherize committed:
simplify definition of eid-table->model + add test

escherize committed:
update tests to take keywords

escherize committed:
improve comment

escherize committed:
generate the eid-table->model map

escherize committed:
delete now-obsolete test

escherize committed:
make it work in oss

escherize committed:
put the resulting response into a key, so we can add more information later

escherize committed:
formatting

escherize committed:
use model names without the model/ prefix as keys

Commit:
Merge branch 'master' into entity-id-translation-service

oisincoveney commented:
In terms of functionality it's working for me! I'll make the FE changes in a different branch so feel free to merge once the BE has approved 😄

oisincoveney commented:
@escherize after doing some work with this, could we actually make these changes:
- could the entities in the map conform to the types found in places like `frontend/src/metabase-types/api/collection.ts` where they're lowercase and conform to the types of the entities:
```
  "card",
  "dataset",
  "metric",
  "dashboard",
  "snippet",
  "collection",
  "indexed-entity",
```

Another question: could we possibly just pass a single entity ID and/or an array of entity IDs, and then return the type and the ID of entity? I do like the API as it us but I think it'll become unwieldy since (from what I am seeing so far) the usage will be confined to single entity IDs for the most part. Something like this:
`EntityId[] => [{type: card, id: 1}, {type: dashboard, id: 123}]`

escherize commented:
> could the entities in the map conform to the types found in places like frontend/src/metabase-types/api/collection.ts where they're lowercase and conform to the types of the entities:

>  ``` 
>  "card"
>  "dataset"
>  "metric"
>  "dashboard"
>  "snippet"
>  "collection"
>  "indexed-entity"
>  ```

I am good with renaming: "card", "dataset", "metric", "dashboard", "snippet", "collection".
 As for `"indexed-entity"`, those don't have entity_ids, so we won't be able to look them up with this endpoint.
 
The complete list of look-uppable models is: `"action"`, `"card"`, `"collection"`, `"dashboard"`, `"dashboard-card"`, `"dashboard-tab"`, `"dimension"`, `"snippet`"), `"permissions-group"`, `"pulse"`, `"pulse-card"`, `"pulse-channel"`, `"segment"`, `"timeline"`, `"user"`.  I see that we have some prior art for referring to models in the search endpoint, and I will extend that a bit so you can refer to them in (I think) an ergonomic way. Some model names are already setup in the search endpoint, and some will be new.

------

>  Do you think you could map the entity ID to an object with a status for each? i.e.:

```
  {
    entity_ids: {
      eid1: {
        id: 1,
        type: "card",
        status: "ok",
      },
      eid2: {
        id: 2,
        type: "model",
        status: "ok",
      },
      eid3: {
        status: "error",
      }
    }
  }
```

Let's do it!

escherize commented:
Had a great huddle with Oisin about this, and I am going to switch up what the response value looks like (see above), and re-use some model names from what we use in search, to work toward a standard for naming models FE <-> BE.

escherize committed:
Creates list of `api/model->db-model`

- update keys for util/entity_id request
- update shape of util/entity_id response
- add test for not-found eids

escherize committed:
Merge branch 'entity-id-translation-service' of github.com:metabase/metabase into entity-id-translation-service

escherize committed:
formatting

escherize committed:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'master' into entity-id-translation-service

johnswanson commented on src/metabase/api/embed/common.clj:
```diff
@@ -488,3 +490,69 @@
+       (map keyword)
+       set))
+
+(def ^{:private true} api-name->model
```
nit: this is the same as  `^:private` right?

escherize replied:
Yeah, this was some cruft I left in when that was a def defining a memoized function. Will fix

johnswanson commented on src/metabase/api/embed/common.clj:
```diff
@@ -488,3 +490,69 @@
+  (let [model (api-name->model api-name) ;; This lookup is safe because we've already validated the api-names
+        eid->id (into {}
+                      (mapv (juxt :entity_id :id)
+                            (t2/select [model :id :entity_id] :entity_id [:in eids])))]
```
Honestly not sure if this is better - IMO your version is _clearer_, but I'm not completely used to all the various options of `t2` yet, so maybe this is more idiomatic?

`(t2/select-fn->fn :entity_id :id [model :id :entity_id] :entity_id [:in eids])`

escherize replied:
Oh, that's pretty cool I wasn't aware of that `t2` function. I'll take it

johnswanson reviewed (approved):
LGTM! Two things too small to even consider nits 😛

escherize committed:
Merge branch 'entity-id-translation-service' of github.com:metabase/metabase into entity-id-translation-service

escherize committed:
Respond to code review feedback

Commit:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'master' into entity-id-translation-service
